cpp/include/cudf/io/parquet.hpp

```diff
@@ -0,0 +1,690 @@
+  std::vector<std::vector<size_type>> _row_groups;
+  /// Number of rows to skip from the start; 0 is none
+  size_type _skip_rows = 0;
+  /// NUmber of rows to read; -1 is all
+  size_type _num_rows = -1;
+
+  /// Whether to store string data as categorical type
+  bool _convert_strings_to_categories = false;
+  /// Whether to use PANDAS metadata to load columns
+  bool _use_pandas_metadata = true;
+  /// Cast timestamp columns to a specific type
+  data_type _timestamp_type{type_id::EMPTY};
+
+  explicit parquet_reader_options(source_info const& src) : _source(src) {}
+
+  friend parquet_reader_options_builder;
+
+ public:
+  /// This is just to please cython
+  explicit parquet_reader_options() = default;
+
+  /**
+   * @brief create parquet_reader_options_builder which will build parquet_reader_options
+   *
+   * @param src source information used to read parquet file
+   * @returns parquet_reader_options_builder
+   */
+  static parquet_reader_options_builder builder(source_info const& src);
+
+  /**
+   * @brief Returns source info
+   */
+  source_info const& source() const { return _source; }
+
+  /**
+   * @breif Returns true/false depending on whether strings should be converted to categories or not
+   */
+  bool convert_strings_to_categories() const { return _convert_strings_to_categories; }
+
+  /**
+   * @breif Returns true/false depending whether to use pandas metadata or not while reading
+   */
+  bool use_pandas_metadata() const { return _use_pandas_metadata; }
+
+  /**
+   * @breif Returns number of rows to skip from the start
+   */
+  size_type skip_rows() const { return _skip_rows; }
+
+  /**
+   * @breif Returns number of rows to read
+   */
+  size_type num_rows() const { return _num_rows; }
+
+  /**
+   * @brief Returns names of column to be read
+   */
+  std::vector<std::string> const& columns() const { return _columns; }
+
+  /**
+   * @brief Returns list of individual row groups to be read
+   */
+  std::vector<std::vector<size_type>> const& row_groups() const { return _row_groups; }
+
+  /**
+   * @brief Returns timestamp_type used to cast timestamp columns
+   */
+  data_type timestamp_type() const { return _timestamp_type; }
+
+  /**
+   * @brief Set column names which needs to be read
+   */
+  void columns(std::vector<std::string> col_names) { _columns = std::move(col_names); }
+
+  /**
+   * @brief Set vector of individual row groups to read
+   */
+  void row_groups(std::vector<std::vector<size_type>> row_grp)
+  {
+    if ((!row_grp.empty()) and ((_skip_rows != 0) or (_num_rows != -1))) {
+      CUDF_FAIL("row_groups can't be set along with skip_rows and num_rows");
+    }
+
+    _row_groups = std::move(row_grp);
+  }
+
+  /*
+   * @brief Set to convert strings to categories
+   */
+  void convert_strings_to_categories(bool val) { _convert_strings_to_categories = val; }
+
+  /*
+   * @brief Set to use pandas metadata to read
+   */
+  void use_pandas_metadata(bool val) { _use_pandas_metadata = val; }
```

Comment:
Getter/setters with the same name and only overloaded based on the presence of function arguments is potentially confusing. Setters should imply mutation in their name with something like `set_`.

Reply:
I felt it would be easier for a developer just to remember the member names without any additional verb or other words. And It is intuitive when you pass a value, you are trying to set it. And, I am not against adding `set_*`, but it would be just be an extra.

Reply:
Looking at lines like 
`options.num_rows(val);`
I do weakly agree with Jake here. 
What bugs me is that we then end up with weird two-verb names like `set_use_pandas_metadata` and `convert_strings_to_categories`, but this is minor.
I don't feel strongly about either option.

@jrhemstad can you please reply to @rgsl888prabhu 's comment above and help make the call?

Reply:
I'm just a casual observer, so don't put too much weight in what I say. That said, I think having getters/setters have the same name and only overload on function parameter is incongruous with a lot of C++ code and will be confusing. I'd advocate for a solution that differentiates them somehow. Maybe @codereport has a good solution.

Reply:
@codereport had also suggested `?` at the end of predicate, but I haven't come across such usage.
https://github.com/rapidsai/cudf/pull/6051#discussion_r475654570

Reply:
How about this: 
```
// parameters use set and get
void set_num_rows(size_type);
size_type get_num_rows();

// options use enable / is_enabled
void enable_use_pandas_metadata(bool enable);
bool is_enabled_use_pandas_metadata();
```
An alternative to `is_enabled_` might be `query_` or use a suffix of `_enabled`: `use_pandas_metadata_enabled()`.

Reply:
This approach looks comparatively better, @jrhemstad @vuule what is your opinion ?

Reply:
I like the `get_`/`set_` option the best of all versions we tried so far (there's probably a commit at which this version of the API can be viewed). 
The `enable_`/`is_enabled_` change has :+1: from me too.